src/main/java/jenkins/plugins/slack/webhook/WebhookEndpoint.java

```diff
@@ -0,0 +1,250 @@
+                StaplerResponse.SC_OK); 
+    
+        String commandText = data.getText();
+        if (commandText == null || commandText.isEmpty())
+            return new JsonResponse(new SlackTextMessage("Invalid command, text field required"),
+                StaplerResponse.SC_OK);
+
+        String triggerWord = data.getTrigger_word();
+        if (triggerWord == null || triggerWord.isEmpty())
+            return new JsonResponse(new SlackTextMessage("Invalid command, trigger_word field required"),
+                StaplerResponse.SC_OK);
+
+        //
+        // TODO: figure out better way of injecting slack user data into handler
+        //
+        this.slackUser = data.getUser_name();
+
+        CommandRouter<SlackTextMessage> router =
+            new CommandRouter<SlackTextMessage>();
+
+        try {
+            router.addRoute("^"+triggerWord+" list projects",
+                triggerWord+" list projects", "Return a list of buildable projects",
+                this,
+                "listProjects")
+            .addRoute("^"+triggerWord+" run ([\\p{L}\\p{N}\\p{ASCII}\\W]+)",
+                triggerWord+" run <project_name>",
+                "Schedule a run for <project_name>",
+                this,
+                "scheduleJob")
+            .addRoute("^"+triggerWord+" get ([\\p{L}\\p{N}\\p{ASCII}\\W]+) #([0-9]+) log",
+                triggerWord+" get <project-name> #<build_number> log",
+                "Return a truncated log for build #<build_number> of <project_name>",
+                this,
+            "getProjectLog");
+
+            SlackTextMessage msg = router.route(commandText);
+
+            return new JsonResponse(msg, StaplerResponse.SC_OK);
+            
+        } catch (RouteNotFoundException ex) {
+
+            LOGGER.warning(ex.getMessage());
+
+            String command = ex.getRouteCommand();
+
+            String response = "*Help:*\n";
+            if (command.split("\\s+").length > 1)
+                response += "`"+command+"` _is an unknown command, try one of the following:_\n\n";
+            else
+                response += "\n";
+
+            for (CommandRouter.Route route : router.getRoutes()) {
+                response += "`"+route.command+"`\n```"+route.commandDescription+"```";
+                response += "\n\n";
+            }
+
+            return new JsonResponse(new SlackTextMessage(response), StaplerResponse.SC_OK);
+
+        } catch (CommandRouterException ex) {
+            LOGGER.warning(ex.getMessage());
+            return new JsonResponse(new SlackTextMessage(ex.getMessage()), StaplerResponse.SC_OK);
+
+        } catch (Exception ex) {
+            LOGGER.warning(ex.getMessage());
+            return new JsonResponse(new SlackTextMessage("An error occured: "+ ex.getMessage()), StaplerResponse.SC_OK);
+        }
+    }
+
+    public SlackTextMessage getProjectLog(String projectName,
+        String buildNumber) {
+
+        ACL.impersonate(ACL.SYSTEM);
```

Comment:
The context should probably be restored, however all methods in the plugin/thread are using the same permission.